### PR TITLE
Fix accessibility issues #1172 #1167  #1171

### LINKED ISSUE #1167: [Deque Analysis] Header's "critical" accessibility issues
> Critical: This issue results in blocked content for people with disabilities, and will definitely prevent them from accessing fundamental features or content. This type of issue puts your organization at risk. Prioritize fixing as soon as possible, within the week if possible. Remediation should be a top priority. 

## Deque Analysis Summary

A total of 9 Critical issues exist in the Header (visible on all pages). They are summarized here:
1. (ID 468402, 468403, 469227, 469228) IDs/Labels must be unique: Two elements exist having each of the ID=`dropdownLang` and ID=`dropdownUser` and ID=`inputEmail` and ID=`inputPassword` . It appears these elements appear twice in the generated HTML code, but only one is visible on the page.  (I'm not sure why this is?)
2. (ID 468406) Searchbox in header has no label.  It should either be given a label, or have the `aria-label` attribute added to describe it.  (Likely the latter is fine, as it doesn't have visible label text by default)
3. (ID 469259) Language dropdown should be a `<select>` instead of a `<ul>` as it is a "listbox".  
    * (ID 469256) Also it needs to be accessible via keyboard (current you can tab to it, but cannot arrow down to select a different language)....this might be fixed by turning this into a `<select>` as well.
4. (ID 468465) Cannot open "All of DSpace" dropdown via keyboard (only triggers on mouse hover).  Likely this should be implemented  similar to the language dropdown or Login dropdown (as those both can be opened by tabbing and clicking Enter)
5. (ID 468473) Search icon button has no `href` value.  It MUST have an `href` value:
   * Fix this issue by providing an HREF value (even if JavaScript event handlers are used) for the link. (Note: Ensure that if JavaScript event handlers are used, they allow the link to be activated with the Enter key in addition to the mouse click.).
   * e.g. `<a href=quickview.html">Quick view</a>`
   * e.g. `<a href="javascript:void(0);">Quick view</a>`

The full list of these issues (which includes links for more info) can be found at this link (requires login): https://axeauditor.dequecloud.com/test-run/0856438a-a19a-11eb-bc31-b7d5be387c86/issues?activeTab=dt-issue&page=0&pageSize=100&sortField=ordinal&sortDir=asc&row=9&filter%5Bseverity%5D=4&filter%5Bcomponent_number%5D=0&filter%5Bpage_number%5D=8

## More Information / Tools

* Consider using [Deque's free Chrome plugin](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US) to check your work

### LINKED ISSUE #1171: [Deque Analysis] Submission / MyDSpace "critical" accessibility issues
> Critical: This issue results in blocked content for people with disabilities, and will definitely prevent them from accessing fundamental features or content. This type of issue puts your organization at risk. Prioritize fixing as soon as possible, within the week if possible. Remediation should be a top priority. 

## Deque Analysis Summary

A total of 28 Critical issues exist in the Submission/MyDSpace pages. They are summarized here:

1. (3 issues) "Elements must only use allowed ARIA attributes."
    * (Issue ID: 469501, 469502) The `aria-expanded` attribute is invalid on `input` fields (it currently appears for dc.type and dc.language.iso fields). See https://www.w3.org/TR/wai-aria-1.1/#aria-expanded  It should be removed from those form fields.
    * ~~(Issue ID: 469503) The `aria-multiline` attribute is invalid on `combobox` role (it currently appears for dc.subject field). See https://www.w3.org/TR/wai-aria-1.1/#aria-multiline  It should be removed.~~ MOVED TO #1272
2. (4 issues) "IDs used in ARIA and labels must be unique." Multiple elements in the submission form currently use ID=`traditionalpageone`, ID=`traditionalpagetwo`, ID=`upload`, ID=`license`.
   * (Issue IDs: 469532, 469533, 469534, 469535)
3. (4 issues) Form/Select elements must have labels.  
    * (Issue ID: 469536) The `dc.type` field is not linked up with its `<label>` tag.
    * ~~(Issue ID: 469538) The `<select>` for the `dc.identifier` field doesn't have a `<label>` or an `aria-label` attribute.~~ MOVED TO #1272
    * (Issue ID: 470007) "Access condition" & "Grant Access" fields in Bitstream embargo are no associated with their visible labels.
    * (Issue ID: 470145) On MyDSpace page, the "Show" selectbox (on left side) is missing / not associated with a label
4. (4 issues) Several Keyboard navigation issues
    * (Issue ID: 469698) "browse" link in the file drag & drop component is inaccessible to keyboard navigation (appears on both Submission & MyDSpace). 
    * ~~(Issue ID: 469927)  Reorder buttons (when field allows multiple values, e.g. Author, Other Title, etc) are inaccessible to keyboard navigation (you cannot tab to them).~~ MOVED TO #1271 
    * (Issue ID: 469979) Arrows above/below date field (to change year/month/day) are inaccessible to keyboard navigation (you cannot tab to them).
    * (Issue ID: 470011) Bitstream embargo fields have a **keyboard trap** (Upload a file -> Edit -> "Access Condition" to "embargo" or "lease" -> Attempt to tab past the "Grant access..." field).  It's not possible to tab past them until you fill them out (Keyboard focus is trapped within the field).  See also https://dequeuniversity.com/class/input-methods2/keyboard-input/keyboard-traps
5. (4 issues) Button does not have a name or role
    * (Issue ID: 469700) "browse" link in the file drag & drop component (`<input type="file">`)  has no "role".  Should be fixed by using a `<button>` tag, or by adding `role=button` attribute.
    * (Issue ID: 470002, 469770) Download, Edit & Remove buttons (red trash can) next to an uploaded file (in "Upload Files" section) all have no name/label.  They likely all need an `aria-label` value.
    * (Issue ID: 470014) Calendar icon/button (on embargo fields) do not have a name/label. 
6. (1 issues) Element's role is incorrect. The "Add more" button on multivalued fields says it has `role=group`, but that's not correct. (Issue ID: 469774)
7. (2 issues) Focus indicator is missing. 
    * (Issue ID: 469873) No focus indicator exists on the form section headings ("Describe", "Describe", "Upload files", "Deposit License").  All elements must have a visual focus indicator when in focus.
        * Common keyboard focus indicator solutions include (choose one): (1) A 2px box around the focused element, (2) A change in the background color of the focused element (3) The addition of an icon, such as an arrow, next to a menu item (4) The addition of a thick line under or next to the focused element (5) A distinctive change to the text of the component such as making it bold and/or underlined (6) Use the default browser focus indicator. NOTE: Since each browser has its own default focus indicator style, check each of the major browsers (Chrome, Firefox, Edge, IE, Safari) to ensure the default focus indicator is visible as you tab through the page.
    * (Issue 470001) No focus indicator exists on File buttons (Download, Edit, Delete...and Save/Cancel after clicking Edit). 
8. (1 Issue) (Issue ID 469984) Group of form controls is missing "group" label.  This occurs on three fields: "Date of issue", "Series/Report No.", "ISSN/Identifiers".  Group labels MUST be associated with the group. To Fix, use either "fieldset/legend" technique or ARIA "group/labelledby" technique. See https://dequeuniversity.com/class/forms2/group-labels/semantic
9. (1 Issue) (Issue ID 469993) Two comboboxes are missing roles/attributes.  Specifically "dc.type" and "dc.language.iso" fields.
    * Two options to fix. Either use `<select>` and `<option>`.  Or, use ARIA design pattern for a combobox, see https://www.w3.org/TR/wai-aria-practices-1.1/#combobox
10. (1 Issue) (Issue ID 470010) Form field with error not identified.  Occurs on the "Grant Access" fields in the Bitstream Embargo section.  When an error occurs it is **not** linked to the field.  Ideally this should occur by using `aria-describedby` attribute to reference a unique ide attribute value on the error message container.  See also https://dequeuniversity.com/class/forms2/form-validation/error-identification
11. (1 Issue) (Issue ID 470062) Section status icons are missing alt text.  When a submission section has a green checkmark, it has all required fields filled. Otherwise it shows an orange exclamation mark.  Both those status icons are missing alt text.
12. (1 issue) (Issue ID 470065) Some elements have incorrect names/labels.
    * "Add more" button has `aria-label="remove button"`
    * Section headers open/close icons (up/down arrows) always say `aria-label="close"`  They probably should say `close section` or `open section` depending on whether the section is open/closed.

The full list of these issues (which includes links for more info) can be found at this link (requires login): https://axeauditor.dequecloud.com/test-run/0856438a-a19a-11eb-bc31-b7d5be387c86/issues?activeTab=dt-issue&page=0&pageSize=100&sortField=ordinal&sortDir=asc&row=9&filter%5Bseverity%5D=4&filter%5Bpage_number%5D=10&filter%5Bpage_number%5D=11

## More Information / Tools

* Consider using [Deque's free Chrome plugin](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US) to check your work

### LINKED ISSUE #1172: [Deque Analysis] Edit Item/Collection "critical" accessibility issues
> Critical: This issue results in blocked content for people with disabilities, and will definitely prevent them from accessing fundamental features or content. This type of issue puts your organization at risk. Prioritize fixing as soon as possible, within the week if possible. Remediation should be a top priority. 

## Deque Analysis Summary

A total of 5 Critical issues exist related to editing Items or Collections. They are summarized here:

1. ~~(Issue ID: 470946) Edit Item, Bitstreams tab: "Content should be in a data table but is not."  The page appears visually as a table, but it is not represented as a `<table>`.~~ MOVED TO #1270 
2. (Issue ID: 471444) Edit Item, Metadata tab: "Label in `<th>` is not associated with form control."  The row and column header cells visually serve as labels for the form controls contained in the table. The form controls are not marked up in a manner that exposes this relationship.
    * Fix this issue by using an aria-labelledby attribute on the <input> to reference the label in another cell. The value of the aria-labelledby attribute is the id attribute value of the visible text label.
    * IMPORTANT: Note that the table headers/labels must be in <span> elements and that it is the <span> element and NOT the <th> element that has the id. If you place the id on the <th> element, some screen readers will not read the labels correctly when you tab through the form elements.
3. ~~(Issue ID: 470936) Edit Item, Bitstreams tab: Button to reorder bitstreams (in a bundle) is not keyboard accessible.~~ MOVED TO #1271 
4. (Issue ID: 470091) Edit Collection/Community, Logo upload is not keyboard accessible
    * NOTE: This same issue appears in uploading files in Submission, see #1171 & it should be solved similarly.
5. (Issue ID: 471077) Edit Collection/Community, Logo/file delete button (red trash icon) does not have a name/label.
    * NOTE: This same issue appears in Submission after uploading files. See #1171 & it should be solved similarly.

The full list of these issues (which includes links for more info) can be found at this link (requires login): https://axeauditor.dequecloud.com/test-run/0856438a-a19a-11eb-bc31-b7d5be387c86/issues?activeTab=dt-issue&page=0&pageSize=100&sortField=ordinal&sortDir=asc&row=9&filter%5Bseverity%5D=4&filter%5Bpage_number%5D=12&filter%5Bpage_number%5D=13&filter%5Bpage_number%5D=14&filter%5Bpage_number%5D=15

## More Information / Tools

* Consider using [Deque's free Chrome plugin](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US) to check your work

### PR DESCRIPTION
**Pull request by atarix83**
## References
* Fixes #1172
* Fixes #1167
* Fixes #1171 

## Description
This issue resolve accessibility problems reported in the issue.

Several issues were moved to new tickets, as they could not be resolved quickly:
* #1270 (from 1172)
* #1271 (from 1172 & 1171)
* #1272 (from 1171)

## Instructions for Reviewers
Check if the issues reported are resolved

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

File: src/app/submission/sections/container/section-container.component.html [hunk 1]
```diff
@@ -12,23 +12,26 @@
...
           <i *ngIf="(sectionRef.hasErrors())" class="fas fa-exclamation-circle text-danger mr-3"
-             aria-hidden="true"></i>
+             aria-hidden="true" title="{{'submission.sections.status.errors.title' | translate}}"></i>
           <i *ngIf="(sectionRef.isValid() | async) && !(sectionRef.hasErrors())" class="fas fa-check-circle text-success mr-3"
-             aria-hidden="true"></i>
-          <a class="close" aria-label="Close">
+             aria-hidden="true" title="{{'submission.sections.status.valid.title' | translate}}"></i>
+          <a class="close"
+             tabindex="0"
+             [attr.aria-label]="(sectionRef.isOpen() ? 'submission.sections.toggle.close' : 'submission.sections.toggle.open') | translate"
```
Comment (tdonohue) [inline src/app/submission/sections/container/section-container.component.html:25]: For some reason Axe scanner is complaining about this `<a>` tag including an `aria-label`...it says that this `aria-label` is not allowed based on the "element's role".  I wonder if there's a way to just turn this into a `<button>`?

**Review by tdonohue — CHANGES_REQUESTED** [review]
@atarix83 : Gave this a test today...While there are some improvements here, there seems to be several issues which were overlooked.  Please make sure you are using [Deque's free Chrome plugin](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US) to check your work on each page...as I found a lot of overlooked "serious" level issues using that tool.  Here's what I see that is outstanding:

1. In #1167, when I scan the header using Deque's Axe plugin, I'm still seeing this error:
    * IDs/Labels must be unique: Two elements exist having each ID=inputEmail and ID=inputPassword (It is now fixed for `ID=dropdownLang` and `ID=dropdownUser`, but still appears for `inputEmail` and `inputPassword`.)
2. Also part of #1167... I'm noticing the two keyboard navigation issues were **not** fixed. 
    * The "All of DSpace" dropdown is inaccessible to keyboard navigation (tabbing to it and pressing Enter). You must click on it with a mouse.
    * The Language dropdown can be opened via keyboard navigation (tabbing to it and pressing Enter), but you **cannot** select a different language via your keyboard once it is opened.  Additionally, as noted in #1167, it was suggested that we may want to change this dropdown into a `<select>` to fix this issue...currently it's a `<ul>`.
3. Part of #1171, on the MyDSpace page, the "Show" selectbox (on left side) is still not connected to a label/name. So, the Axe Tools still show a "Select element must have an accessible name" error.  This is part of issue 3 in that ticket.
4. Part of #1171 on the Submission page, the "Delete" icon (next to multi-valued values) is showing up as having an invalid ARIA attribute.  The Deque Axe tools say this cannot have an `aria-label` as it's a `<span>`.  This might be fixed by turning this into a `<button>` 
5. Part of #1171 on the Submission page, the "Subject Keywords" field has an invalid ARIA attribute of `aria-expanded`.  This isn't valid on an `<input>` as that's not a collapsible field.
6. Part of #1171 on the Submission page...still seeing an error that the `dc.type` field is *unlabelled*.  It looks like it has a `aria-labelledby="label_dc_type"`, but there's no label that matches that value.
7. Part of #1171 on the Submission page.  the Identifiers selectbox (ISSN, ISBN, etc) also has no label.
8. Part of #1172 (number 5 in that ticket) on Edit Collection page, when logo is uploaded, the red trash icon has no label or text.
9. Part of #1172 (number 3 in that ticket) the button to reorder bitstreams in Item Edit is not keyboard accessible.  I can tab to it, but clicking "Enter" does nothing.  If this is hard to achieve we could move it to a later ticket.

There's one inline comment as well (see below)

File: src/app/submission/sections/container/section-container.component.html (hunk 1 above)
Comment (atarix83) [reply src/app/submission/sections/container/section-container.component.html:26]: fixed by adding a role button to <a> tag

**Comment by atarix83** [comment]
@tdonohue 

I fixed all except for the following : 

> Part of [Deque Analysis] Submission / MyDSpace "critical" accessibility issues #1171 on the Submission page, the "Subject Keywords" field has an invalid ARIA attribute of aria-expanded. This isn't valid on an <input> as that's not a collapsible field.

the aria is added by the [ngbTypeahead component,](https://ng-bootstrap.github.io/#/components/typeahead/examples) I can't avoid it


> Part of [Deque Analysis] Submission / MyDSpace "critical" accessibility issues #1171 on the Submission page. the Identifiers selectbox (ISSN, ISBN, etc) also has no label.

I can't find a solution because that element is added by the form library

> Part of [Deque Analysis] Edit Item/Collection "critical" accessibility issues #1172 (number 3 in that ticket) the button to reorder bitstreams in Item Edit is not keyboard accessible. I can tab to it, but clicking "Enter" does nothing. If this is hard to achieve we could move it to a later ticket.

this is hard to achieve, it should be done in a specific ticket

**Comment by tdonohue** [comment]
@atarix83 : It looks like recent updates have resulted in a test failure in this PR.  So, while I can go ahead and review it, we'll need to get the test fixed as well obviously. Thanks!

**Review by tdonohue — CHANGES_REQUESTED** [review]
@atarix83 : It's looking better now, but after testing, I still see a few issues remaining:

1. #1167 appears to be fixed EXCEPT FOR ONE ISSUE
   * (ID 469256) Language Dropdown needs to be accessible via keyboard (current you can tab to it, but cannot arrow down to select a different language)....this might be fixed by turning this into a <select> as well.
   * The behavior is improved. I can now tab to a different language, but if I click "Enter" nothing happens. There is no way to select a different language without using your mouse.  If somehow this is too hard to achieve, we can split this out.
2. #1172 appears to be fixed EXCEPT FOR ONE Issue
   * (Issue ID: 471077) Edit Collection/Community, Logo/file delete button (red trash icon) does not have a name/label. (To reproduce, you have to first upload a logo...then scan the page when you see the red trash can icon) 
3. #1171 appears to be fixed (except where you noted)

So, this is very close now.  In the meantime, I'll work on moving the "known issues" (ones you've said you cannot fix quickly) into separate tickets so that they can be rescheduled for 7.1.

**Comment by atarix83** [comment]
@tdonohue 
> [[Deque Analysis] Header's "critical" accessibility issues #1167](https://github.com/DSpace/dspace-angular/issues/1167) appears to be fixed EXCEPT FOR ONE ISSUE
> (ID 469256) Language Dropdown needs to be accessible via keyboard (current you can tab to it, but cannot arrow down to select a different language)....this might be fixed by turning this into a `<select>` as well.
> The behavior is improved. I can now tab to a different language, but if I click "Enter" nothing happens. There is no way to select a different language without using your mouse.  If somehow this is too hard to achieve, we can split this out.

I've fixed it, now it's possible to select a language entry by pressing "enter" 



> [[Deque Analysis] Edit Item/Collection "critical" accessibility issues #1172](https://github.com/DSpace/dspace-angular/issues/1172) appears to be fixed EXCEPT FOR ONE Issue
> (ID: 471077) Edit Collection/Community, Logo/file delete button (red trash icon) does not have a name/label. (To reproduce, you have to first upload a logo...then scan the page when you see the red trash can icon)

Both Edit Item/Collection and Logo/file delete button already have an aria-label attribute, I don't kown how to fix otherwise

File: src/app/shared/comcol-forms/comcol-form/comcol-form.component.html [hunk 2]
```diff
@@ -12,11 +12,13 @@
           <div class="col-4 d-inline-block">
             <div *ngIf="logo" class="btn-group btn-group-sm float-right" role="group">
               <button *ngIf="!markLogoForDeletion" type="button" class="btn btn-danger"
+                      [attr.aria-label]="(type.value + '.edit.logo.delete.title' | translate)"
```
Comment (tdonohue) [inline src/app/shared/comcol-forms/comcol-form/comcol-form.component.html:15]: This line & line 21 in this same file both are not working as expected.  The `aria-label` never appears on the `<button>` when you inspect the final page.  I also notice that you already had a `title` attribute on the `<button>`, which I think would **fix this problem** without the need for an `aria-label`.  But, that `title` attribute also never appears in the final code.  Maybe there's a bug in the `translate` line?

File: src/app/shared/comcol-forms/comcol-form/comcol-form.component.html [hunk 3]
```diff
@@ -12,11 +12,13 @@
...
                       title="{{type.value + '.edit.logo.delete.title' | translate}}"
                       (click)="deleteLogo()">
                 <i class="fas fa-trash" aria-hidden="true"></i>
               </button>
               <button *ngIf="markLogoForDeletion" type="button" class="btn btn-warning"
+                      [attr.aria-label]="(type.value + '.edit.logo.delete-undo.title' | translate)"
```
Comment (tdonohue) [inline src/app/shared/comcol-forms/comcol-form/comcol-form.component.html:21]: Same issue occurs here.  Again, you already have a `title` on the next line, but neither the `title` nor the `aria-label` appear in the final source.

**Review by tdonohue — APPROVED** [review]
👍 @atarix83 : This looks "good enough".  However, I will note that this issue still is definitely not fixed:
> [Deque Analysis] Edit Item/Collection "critical" accessibility issues #1172 appears to be fixed EXCEPT FOR ONE Issue
(ID: 471077) Edit Collection/Community, Logo/file delete button (red trash icon) does not have a name/label. (To reproduce, you have to first upload a logo...then scan the page when you see the red trash can icon)

It looks like you *attempted* to fix the issue, but the "aria-label" attribute you added never appears on the `<button>` tag when you inspect the page in the browser.  I also noticed that same `<button>` doesn't have a `title` attribute even though you attempted to add that as well.  So, I think there's a bug in the code you added there (see inline comments below as well)

If you have a chance to glance at that again, it might be very quick to fix.  Otherwise, I'll merge this on/by Monday, and we'll just have to solve that bug at a later time.  Overall, this PR looks good though and looks (to me) like it fixes everything else that it is supposed to. Thanks!

File: src/app/shared/comcol-forms/comcol-form/comcol-form.component.html (hunk 2 above)
Comment (atarix83) [reply src/app/shared/comcol-forms/comcol-form/comcol-form.component.html:15]: @tdonohue 
I had a look and if I inspect the html code tith the devtools I can see both the title and the aria.label property
Schermata da 2021-07-19 16-23-14

Moreover if I scan the page with the "axe DevTools" I don't get the accebility error, while If I remove bothe title and aria.label, I get the accessibility error "Buttons must have discernible text".

Comment (tdonohue) [reply src/app/shared/comcol-forms/comcol-form/comcol-form.component.html:15]: @atarix83 : I checked again, and it looks like we are testing slightly different scenarios. You are correct, the `title` and `aria.label` property **do appear after clicking Save**.  But, they are **missing after the initial upload**.  Here's what I see, step by step..

1. Login as an Admin
2. Go to a Collection _which doesn't have a logo_
3. Edit the Collection. Drag & drop a logo.  Trash can icon appears (DO NOT CLICK SAVE YET)
4. Inspect the page... you'll see the Trash can `<button>` has no `title` nor `aria-label`
5. Now click Save. You go back to the Collection page
6. Click Edit.  Inspect the page.  Now, the Trash can `<button>` has both the `title` and the `aria-label`.

So, it seems I was testing immediately after upload, but you are correct the label & title do appear after saving.   Ideally, the attributes should appear in **both scenarios**.  But, if we cannot figure out a way to easily do so, this PR is likely "good enough" for now.

Comment (atarix83) [reply src/app/shared/comcol-forms/comcol-form/comcol-form.component.html]: @tdonohue yeah we're testing different scenarios. Now it should be fixed. I've removed unnecessary `aria-label`

Comment (tdonohue) [reply src/app/shared/comcol-forms/comcol-form/comcol-form.component.html]: That fixed it! Thanks @atarix83 !  I'll go ahead and merge this now.

**Comment by tdonohue** [comment]
👍 Looks good to me now.  All unresolved issues have been moved to separate tickets (see links in updated description above)

**Pull request merged**